### PR TITLE
Add additional metrics

### PR DESCRIPTION
**Pull request by mslipper**
Continues addressing https://github.com/cosmos/cosmos-sdk/issues/2169.

* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: mempool/metrics.go [hunk 1]
```diff
@@ -3,32 +3,62 @@ package mempool
...
+			Subsystem: MetricsSubsytem,
 			Name:      "size",
 			Help:      "Size of the mempool (number of uncommitted transactions).",
 		}, []string{}),
+		TxSizeBytes: prometheus.NewHistogramFrom(stdprometheus.HistogramOpts{
+			Namespace: namespace,
+			Subsystem: MetricsSubsytem,
+			Name:      "tx_size_bytes",
+			Help:      "Transaction sizes in bytes.",
+			Buckets:   stdprometheus.ExponentialBuckets(1, 2, 10),
```
Comment (melekes) [inline mempool/metrics.go]: what's the last bucket - 1024 bytes? some of our users have 100KB transactions. the max size for a transaction right now is 21MB (block size)

File: mempool/metrics.go [hunk 2]
```diff
@@ -3,32 +3,62 @@ package mempool
...
+
 // Metrics contains metrics exposed by this package.
 // see MetricsProvider for descriptions.
 type Metrics struct {
 	// Size of the mempool.
 	Size metrics.Gauge
+	// Histogram of transaction sizes, in bytes.
+	TxSizeBytes metrics.Histogram
+	// Number of failed transactions.
+	FailedTxs metrics.Gauge
```
Comment (melekes) [inline mempool/metrics.go]: why gauge if it's only growing?

File: state/metrics.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	BlockProcessingTime metrics.Histogram
+}
+
+func PrometheusMetrics(namespace string) *Metrics {
+	return &Metrics{
+		BlockProcessingTime: prometheus.NewHistogramFrom(stdprometheus.HistogramOpts{
+			Namespace: namespace,
+			Subsystem: MetricsSubsystem,
+			Name:      "block_processing_time",
+			Help:      "Time between BeginBlock and EndBlock.",
```
Comment (melekes) [inline state/metrics.go]: unit? seconds? ms?

File: mempool/metrics.go [hunk 4]
```diff
@@ -36,9 +36,9 @@ func PrometheusMetrics(namespace string) *Metrics {
 			Subsystem: MetricsSubsytem,
 			Name:      "tx_size_bytes",
 			Help:      "Transaction sizes in bytes.",
-			Buckets:   stdprometheus.ExponentialBuckets(1, 2, 10),
+			Buckets:   stdprometheus.ExponentialBuckets(1, 2, 24),
```
Comment (melekes) [inline mempool/metrics.go]: Do you think the curve should be more steep with less number of buckets? (1, 3, 17) => 9. 27, 81, 243, 729, 2KB, 6KB, 19KB, 59KB, 177KB, 531KB, 1.5MB, 4,7MB

Comment (mslipper) [reply mempool/metrics.go]: Sure - I'm unfamiliar with the best bucketing strategy, so will happily defer to you.

**Comment by melekes** [comment]
We also need to update docs! https://tendermint.com/docs/tendermint-core/metrics.html#list-of-available-metrics

**Comment by mslipper** [comment]
@melekes docs should be good to go :)

File: consensus/reactor.go [hunk 5]
```diff
@@ -827,12 +827,12 @@ func (conR *ConsensusReactor) peerStatsRoutine() {
 			case *VoteMessage:
 				if numVotes := ps.RecordVote(); numVotes%votesToContributeToBecomeGoodPeer == 0 {
 					conR.Switch.MarkPeerAsGood(peer)
-				}
+	}
```
Comment (xla) [inline consensus/reactor.go]: This reformatting seems incorrect.

File: consensus/reactor.go [hunk 6]
```diff
@@ -827,12 +827,12 @@ func (conR *ConsensusReactor) peerStatsRoutine() {
...
 			case *BlockPartMessage:
 				if numParts := ps.RecordBlockPart(); numParts%blocksToContributeToBecomeGoodPeer == 0 {
 					conR.Switch.MarkPeerAsGood(peer)
-				}
 			}
+}
```
Comment (xla) [inline consensus/reactor.go]: Incorrect reformatting.

File: docs/tendermint-core/metrics.md [hunk 7]
```diff
@@ -8,35 +8,45 @@ This functionality is disabled by default.
...
+p2p\_peer\_receive\_bytes\_total|counter|on dev|peer\_id|number of bytes received from a given peer
+p2p\_peer\_send\_bytes\_total|counter|on dev|peer\_id|number of bytes sent to a given peer
+p2p\_peer\_pending\_send\_bytes|gauge|on dev|peer\_id|number of pending bytes to be sent to a given peer
+p2p\_num\_txs |gauge |on dev|peer\_id |number of transactions submitted by each peer\_id
+p2p\_pending\_send\_bytes |gauge |on dev|peer\_id |amount of data pending to be sent to peer
+mempool\_size|Gauge|0.21.0| |Number of uncommitted transactions
+mempool\_tx\_size\_bytes |histogram |on dev| |transaction sizes in bytes
+mempool\_failed\_txs |counter|on dev| |number of failed transactions
+mempool\_recheck\_times |counter|on dev| |number of transactions rechecked in the mempool
+state\_block\_processing\_time|histogram|on dev| |time between BeginBlock and EndBlock in ms
```
Comment (xla) [inline docs/tendermint-core/metrics.md]: This reformatting makes it harder to read the content. Let's revert.

File: state/execution.go [hunk 8]
```diff
@@ -418,7 +439,7 @@ func fireEvents(logger log.Logger, eventBus types.BlockEventPublisher, block *ty
 // ExecCommitBlock executes and commits a block on the proxyApp without validating or mutating the state.
 // It returns the application root hash (result of abci.Commit).
 func ExecCommitBlock(appConnConsensus proxy.AppConnConsensus, block *types.Block,
-	logger log.Logger, lastValSet *types.ValidatorSet, stateDB dbm.DB) ([]byte, error) {
+		logger log.Logger, lastValSet *types.ValidatorSet, stateDB dbm.DB) ([]byte, error) {
```
Comment (xla) [inline state/execution.go]: If we reformat this than one param per line wtih trailing comma and newline after the last param.

File: state/execution.go [hunk 9]
```diff
@@ -334,7 +355,7 @@ func updateValidators(currentSet *types.ValidatorSet, abciUpdates []abci.Validat
 
 // updateState returns a new State updated according to the header and responses.
 func updateState(state State, blockID types.BlockID, header *types.Header,
-	abciResponses *ABCIResponses) (State, error) {
+		abciResponses *ABCIResponses) (State, error) {
```
Comment (xla) [inline state/execution.go]: If we reformat this than one param per line wtih trailing comma and newline after the last param.

File: state/execution.go [hunk 10]
```diff
@@ -177,7 +198,7 @@ func (blockExec *BlockExecutor) Commit(
 // Executes block's transactions on proxyAppConn.
 // Returns a list of transaction results and updates to the validator set
 func execBlockOnProxyApp(logger log.Logger, proxyAppConn proxy.AppConnConsensus,
-	block *types.Block, lastValSet *types.ValidatorSet, stateDB dbm.DB) (*ABCIResponses, error) {
+		block *types.Block, lastValSet *types.ValidatorSet, stateDB dbm.DB) (*ABCIResponses, error) {
```
Comment (xla) [inline state/execution.go]: If we reformat this than one param per line wtih trailing comma and newline after the last param.

File: node/node.go [hunk 11]
```diff
@@ -289,7 +290,7 @@ func NewNode(config *cfg.Config,
 
 	blockExecLogger := logger.With("module", "state")
 	// make block executor for consensus and blockchain reactors to execute blocks
-	blockExec := sm.NewBlockExecutor(stateDB, blockExecLogger, proxyApp.Consensus(), mempool, evidencePool)
+	blockExec := sm.NewBlockExecutor(stateDB, blockExecLogger, proxyApp.Consensus(), mempool, evidencePool, sm.BlockExecutorWithMetrics(smMetrics))
```
Comment (xla) [inline node/node.go]: This line should be broken up, it grew quit big.

**Comment by jackzampolin** [comment]
I think these comments are addressed @xla! Looks like this is ready to merge.

**Pull request merged**